core/pkg/artifacts/saver.go

```diff
@@ -301,6 +328,119 @@ func newUploadTask(fileInfo serverFileResponse, localPath string) *filetransfer.
 	}
 }
 
+const (
+	S3MinMultiUploadSize = 2 << 30 // 2 GiB, the threshold we've chosen to switch to multipart
+	S3MaxMultiUploadSize = 5 << 40 // 5 TiB, maximum possible object size
+	S3MaxParts           = 10000
+)
+
+func multiPartRequest(path string) ([]gql.UploadPartsInput, error) {
+	fileInfo, err := os.Stat(path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get file size for path %s: %w", path, err)
+	}
+	fileSize := fileInfo.Size()
+
+	if fileSize < S3MinMultiUploadSize {
+		// We don't need to use multipart for small files.
+		return nil, nil
+	}
+	if fileSize > S3MaxMultiUploadSize {
+		return nil, fmt.Errorf("file size exceeds maximum S3 object size: %v", fileSize)
+	}
+
+	file, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	partsInfo := []gql.UploadPartsInput{}
+	partNumber := int64(0)
+	buffer := make([]byte, getChunkSize(fileSize))
+	for {
+		bytesRead, err := file.Read(buffer)
+		if err != nil && err != io.EOF {
+			return nil, err
+		}
+		if bytesRead == 0 {
+			break
+		}
+		partsInfo = append(partsInfo, gql.UploadPartsInput{
+			PartNumber: partNumber,
+			HexMD5:     utils.ComputeHexMD5(buffer[:bytesRead]),
+		})
+		partNumber++
+	}
+	return partsInfo, nil
+}
+
+func (as *ArtifactSaver) uploadMultipart(path string, fileInfo serverFileResponse,
+	partData []gql.UploadPartsInput, doneChan chan<- *filetransfer.Task,
+) {
```

Comment:
formatting nit (`gofmt` leaves this up to the author): when the parameters list doesn't fit on one ~80 character line, put each parameter on its own line:

```suggestion
func (as *ArtifactSaver) uploadMultipart(
	path string,
	fileInfo serverFileResponse,
	partData []gql.UploadPartsInput,
	doneChan chan<- *filetransfer.Task,
) {
```

vertical alignment generally makes it easier to scan through the code

Reply:
Thanks, I'm used to that with Python/Black, I like it with one parameter per line better anyway